app/palette/csrf_storage.py

```diff
@@ -114,43 +178,199 @@ def delete_old_token():
+        self.run_sql("DELETE FROM {0} WHERE user_name = '{1}'", self.table_name, user_name)
+
+    def failed_attempt(self, user_name):
+        """ Increments the failed attempt counter for the user """
+
+        # Check if we have the user
+
+        attempt_count =  self._get_attempt_count(user_name)
+
+        if attempt_count > 0:
+            self.run_sql("UPDATE {0} SET attempt_count={1} WHERE user_name='{2}'",
+                         self.table_name, attempt_count + 1, user_name)
+        else:
+            # Create a new record
+            self.run_sql("INSERT INTO {0} (user_name, attempt_count) VALUES ('{1}', 1)",
+                         self.table_name, user_name)
+
+
+
+    def is_user_locked_out(self, user_name):
+        """ Returns True if the user is locked out because of too many login attempts """
+        return self._get_attempt_count(user_name) > self.max_attempts
+
+
+    def _get_attempt_count(self, user_name):
+        """ Returns the number of attempts the user has tried to login unsuccesfully (since they were last cleared) """
+        return self.run_single_value(0, "SELECT attempt_count FROM {0} WHERE user_name = '{1}'",
+                                     self.table_name, user_name)
+
+
+    def _table_sql(self):
+        """ The table columns for the table defition to be created """
+        return """
+            user_name VARCHAR(64) NOT NULL PRIMARY KEY,
+            attempt_count INT NOT NULL
+        """
+
+
+class EmptyUserLockoutStorage():
+    """ Interface to implement for User Lockout backing storage """
+
+    def successful_attempt(self, user_name):
+        """ Clears the record of the users failed login attempts after a successful login """
+        raise NotImplementedError("successful_attempt() not implemented")
+
+    def failed_attempt(self, user_name):
+        """ Increments the failed attempt counter for the user """
+        raise NotImplementedError("failed_attempt() not implemented")
+
+    def is_user_locked_out(self, user_name):
+        """ Returns True if the user is locked out because of too many login attempts """
+        raise NotImplementedError("is_user_locked_out() not implemented")
+
+
+# TODO: move this import
+from cgi import parse_qs
+
+def parse_request_body(env):
+    try:
+        # CONTENT_LENGTH may be empty or missing
+        request_body_size = int(env.get('CONTENT_LENGTH', 0))
+        # Read the request body
+        request_body = env['wsgi.input'].read(request_body_size)
+        # Attempt to parse the request body as query string
+        return parse_qs(request_body)
+    except ValueError:
+        # No body is present, or not in query-string format
+        # so we return an empty
+        return {}
+
+
+class UserLockoutMiddleware(LoggingComponent):
+    def __init__(self, app,
+                 storage=EmptyUserLockoutStorage(),
+                 logger=logging,
+                 user_name_param='username'):
+        """ Creates a new CSRF middleware.
```

Comment:
Params of __init__ do not match docstring

Reply:
fixed in e3752c3